File: backend/services/github_service.py
"""
GitHub API service for repository management and webhook setup
"""
import os
import httpx
import logging
from typing import Dict, Any, List, Optional

logger = logging.getLogger(__name__)

class GitHubService:
    def __init__(self):
        self.token = os.getenv("GITHUB_TOKEN")
        self.base_url = "https://api.github.com"
        
        if not self.token:
            raise ValueError("GITHUB_TOKEN environment variable is required")
    
    async def validate_repository(self, owner: str, repo: str) -> bool:
        """Validate that a GitHub repository exists and is accessible"""
        try:
            async with httpx.AsyncClient() as client:
                headers = {
                    "Authorization": f"token {self.token}",
                    "Accept": "application/vnd.github.v3+json"
                }
                
                response = await client.get(
                    f"{self.base_url}/repos/{owner}/{repo}",
                    headers=headers,
                    timeout=30.0
                )
                
                return response.status_code == 200
                
        except Exception as e:
            logger.error(f"Error validating repository {owner}/{repo}: {e}")
            return False
    
    async def get_user_repositories(self) -> List[Dict[str, Any]]:
        """Get repositories accessible to the authenticated user"""
        try:
            async with httpx.AsyncClient() as client:
                headers = {
                    "Authorization": f"token {self.token}",
                    "Accept": "application/vnd.github.v3+json"
                }
                
                # Get user repos
                response = await client.get(
                    f"{self.base_url}/user/repos",
                    headers=headers,
                    params={
                        "sort": "updated",
                        "per_page": 100,
                        "type": "all"
                    },
                    timeout=30.0
                )
                
                response.raise_for_status()
                repos = response.json()
                
                # Also get organization repos
                org_repos = []
                try:
                    orgs_response = await client.get(
                        f"{self.base_url}/user/orgs",
                        headers=headers,
                        timeout=30.0
                    )
                    
                    if orgs_response.status_code == 200:
                        orgs = orgs_response.json()
                        
                        for org in orgs:
                            org_repos_response = await client.get(
                                f"{self.base_url}/orgs/{org['login']}/repos",
                                headers=headers,
                                params={"per_page": 100},
                                timeout=30.0
                            )
                            
                            if org_repos_response.status_code == 200:
                                org_repos.extend(org_repos_response.json())
                                
                except Exception as e:
                    logger.warning(f"Failed to get organization repos: {e}")
                
                # Combine and deduplicate
                all_repos = repos + org_repos
                seen = set()
                unique_repos = []
                
                for repo in all_repos:
                    if repo['id'] not in seen:
                        seen.add(repo['id'])
                        unique_repos.append({
                            'id': repo['id'],
                            'name': repo['name'],
                            'full_name': repo['full_name'],
                            'owner': repo['owner'],
                            'description': repo.get('description'),
                            'private': repo['private'],
                            'updated_at': repo['updated_at']
                        })
                
                # Sort by updated date
                unique_repos.sort(key=lambda x: x['updated_at'], reverse=True)
                
                logger.info(f"Retrieved {len(unique_repos)} repositories")
                return unique_repos
                
        except httpx.HTTPError as e:
            logger.error(f"HTTP error getting repositories: {e}")
            raise Exception(f"Failed to get repositories: {e}")
        except Exception as e:
            logger.error(f"Error getting repositories: {e}")
            raise
    
    async def setup_webhook(self, owner: str, repo: str, webhook_url: str) -> Dict[str, Any]:
        """Set up a webhook for the repository"""
        try:
            async with httpx.AsyncClient() as client:
                headers = {
                    "Authorization": f"token {self.token}",
                    "Accept": "application/vnd.github.v3+json",
                    "Content-Type": "application/json"
                }
                
                # Check if webhook already exists
                existing_webhooks = await self.get_webhooks(owner, repo)
                for webhook in existing_webhooks:
                    if webhook.get('config', {}).get('url') == webhook_url:
                        logger.info(f"Webhook already exists for {owner}/{repo}")
                        return webhook
                
                # Create new webhook
                payload = {
                    "name": "web",
                    "active": True,
                    "events": [
                        "pull_request",
                        "push",
                        "pull_request_review",
                        "issue_comment"
                    ],
                    "config": {
                        "url": webhook_url,
                        "content_type": "json",
                        "insecure_ssl": "0"
                    }
                }
                
                response = await client.post(
                    f"{self.base_url}/repos/{owner}/{repo}/hooks",
                    headers=headers,
                    json=payload,
                    timeout=30.0
                )
                
                response.raise_for_status()
                result = response.json()
                
                logger.info(f"Created webhook for {owner}/{repo}: {result.get('id')}")
                return result
                
        except httpx.HTTPError as e:
            logger.error(f"HTTP error setting up webhook: {e}")
            raise Exception(f"Failed to setup webhook: {e}")
        except Exception as e:
            logger.error(f"Error setting up webhook: {e}")
            raise
    
    async def get_webhooks(self, owner: str, repo: str) -> List[Dict[str, Any]]:
        """Get existing webhooks for a repository"""
        try:
            async with httpx.AsyncClient() as client:
                headers = {
                    "Authorization": f"token {self.token}",
                    "Accept": "application/vnd.github.v3+json"
                }
                
                response = await client.get(
                    f"{self.base_url}/repos/{owner}/{repo}/hooks",
                    headers=headers,
                    timeout=30.0
                )
                
                response.raise_for_status()
                return response.json()
                
        except httpx.HTTPError as e:
            logger.error(f"HTTP error getting webhooks: {e}")
            return []
        except Exception as e:
            logger.error(f"Error getting webhooks: {e}")
            return []
    
    async def get_pull_request(self, owner: str, repo: str, pr_number: int) -> Optional[Dict[str, Any]]:
        """Get pull request information"""
        try:
            async with httpx.AsyncClient() as client:
                headers = {
                    "Authorization": f"token {self.token}",
                    "Accept": "application/vnd.github.v3+json"
                }
                
                response = await client.get(
                    f"{self.base_url}/repos/{owner}/{repo}/pulls/{pr_number}",
                    headers=headers,
                    timeout=30.0
                )
                
                response.raise_for_status()
                return response.json()
                
        except httpx.HTTPError as e:
            logger.error(f"HTTP error getting pull request: {e}")
            return None
        except Exception as e:
            logger.error(f"Error getting pull request: {e}")
            return None
    
    async def merge_pull_request(self, owner: str, repo: str, pr_number: int, 
                                commit_title: str = "", commit_message: str = "") -> Dict[str, Any]:
        """Merge a pull request"""
        try:
            async with httpx.AsyncClient() as client:
                headers = {
                    "Authorization": f"token {self.token}",
                    "Accept": "application/vnd.github.v3+json",
                    "Content-Type": "application/json"
                }
                
                payload = {
                    "commit_title": commit_title or f"Merge pull request #{pr_number}",
                    "commit_message": commit_message,
                    "merge_method": "merge"
                }
                
                response = await client.put(
                    f"{self.base_url}/repos/{owner}/{repo}/pulls/{pr_number}/merge",
                    headers=headers,
                    json=payload,
                    timeout=30.0
                )
                
                response.raise_for_status()
                result = response.json()
                
                logger.info(f"Merged pull request #{pr_number} in {owner}/{repo}")
                return result
                
        except httpx.HTTPError as e:
            logger.error(f"HTTP error merging pull request: {e}")
            raise Exception(f"Failed to merge pull request: {e}")
        except Exception as e:
            logger.error(f"Error merging pull request: {e}")
            raise
    
    async def create_issue_comment(self, owner: str, repo: str, issue_number: int, body: str) -> Dict[str, Any]:
        """Create a comment on an issue or pull request"""
        try:
            async with httpx.AsyncClient() as client:
                headers = {
                    "Authorization": f"token {self.token}",
                    "Accept": "application/vnd.github.v3+json",
                    "Content-Type": "application/json"
                }
                
                payload = {"body": body}
                
                response = await client.post(
                    f"{self.base_url}/repos/{owner}/{repo}/issues/{issue_number}/comments",
                    headers=headers,
                    json=payload,
                    timeout=30.0
                )
                
                response.raise_for_status()
                result = response.json()
                
                logger.info(f"Created comment on issue #{issue_number} in {owner}/{repo}")
                return result
                
        except httpx.HTTPError as e:
            logger.error(f"HTTP error creating issue comment: {e}")
            raise Exception(f"Failed to create issue comment: {e}")
        except Exception as e:
            logger.error(f"Error creating issue comment: {e}")
            raise

File: frontend/src/services/githubService.ts
/**
 * GitHub API service for repository management
 */

// ============================================================================
// GITHUB SERVICE CLASS
// ============================================================================

class GitHubService {
  private token: string;
  private baseUrl = 'https://api.github.com';

  constructor() {
    this.token = process.env.REACT_APP_GITHUB_TOKEN || '';
    if (!this.token) {
      console.warn('GitHub token not configured. Some features may not work.');
    }
  }

  // ========================================================================
  // AUTHENTICATION
  // ========================================================================

  /**
   * Set GitHub token
   */
  setToken(token: string): void {
    this.token = token;
  }

  /**
   * Check if authenticated
   */
  isAuthenticated(): boolean {
    return !!this.token;
  }

  // ========================================================================
  // API HELPERS
  // ========================================================================

  /**
   * Make authenticated request to GitHub API
   */
  private async makeRequest(endpoint: string, options: RequestInit = {}): Promise<any> {
    if (!this.token) {
      throw new Error('GitHub token not configured');
    }

    const url = `${this.baseUrl}${endpoint}`;
    const headers = {
      'Authorization': `Bearer ${this.token}`,
      'Accept': 'application/vnd.github.v3+json',
      'Content-Type': 'application/json',
      ...options.headers
    };

    try {
      const response = await fetch(url, {
        ...options,
        headers
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(`GitHub API error: ${response.status} - ${errorData.message || response.statusText}`);
      }

      return response.json();
    } catch (error) {
      console.error(`GitHub API request failed: ${endpoint}`, error);
      throw error;
    }
  }

  // ========================================================================
  // USER OPERATIONS
  // ========================================================================

  /**
   * Get current authenticated user
   */
  async getCurrentUser(): Promise<any> {
    return this.makeRequest('/user');
  }

  /**
   * Get user repositories
   */
  async getUserRepositories(username?: string, type: 'all' | 'owner' | 'member' = 'owner'): Promise<any[]> {
    const endpoint = username ? `/users/${username}/repos` : '/user/repos';
    const params = new URLSearchParams({
      type,
      sort: 'updated',
      per_page: '100'
    });

    return this.makeRequest(`${endpoint}?${params}`);
  }

  /**
   * Get organization repositories
   */
  async getOrganizationRepositories(org: string): Promise<any[]> {
    const params = new URLSearchParams({
      sort: 'updated',
      per_page: '100'
    });

    return this.makeRequest(`/orgs/${org}/repos?${params}`);
  }

  // ========================================================================
  // REPOSITORY OPERATIONS
  // ========================================================================

  /**
   * Get repository details
   */
  async getRepository(fullName: string): Promise<any> {
    return this.makeRequest(`/repos/${fullName}`);
  }

  /**
   * Get repository branches
   */
  async getBranches(fullName: string): Promise<any[]> {
    return this.makeRequest(`/repos/${fullName}/branches`);
  }

  /**
   * Get repository contents
   */
  async getContents(fullName: string, path: string = '', ref?: string): Promise<any> {
    const params = new URLSearchParams();
    if (ref) params.append('ref', ref);
    
    const query = params.toString() ? `?${params}` : '';
    return this.makeRequest(`/repos/${fullName}/contents/${path}${query}`);
  }

  /**
   * Search repositories
   */
  async searchRepositories(query: string, sort: 'stars' | 'forks' | 'updated' = 'updated'): Promise<any> {
    const params = new URLSearchParams({
      q: query,
      sort,
      order: 'desc',
      per_page: '50'
    });

    return this.makeRequest(`/search/repositories?${params}`);
  }

  // ========================================================================
  // PULL REQUEST OPERATIONS
  // ========================================================================

  /**
   * Get pull requests
   */
  async getPullRequests(
    fullName: string, 
    state: 'open' | 'closed' | 'all' = 'open',
    sort: 'created' | 'updated' | 'popularity' = 'updated'
  ): Promise<any[]> {
    const params = new URLSearchParams({
      state,
      sort,
      direction: 'desc',
      per_page: '50'
    });

    return this.makeRequest(`/repos/${fullName}/pulls?${params}`);
  }

  /**
   * Get pull request details
   */
  async getPullRequest(fullName: string, prNumber: number): Promise<any> {
    return this.makeRequest(`/repos/${fullName}/pulls/${prNumber}`);
  }

  /**
   * Get pull request files
   */
  async getPullRequestFiles(fullName: string, prNumber: number): Promise<any[]> {
    return this.makeRequest(`/repos/${fullName}/pulls/${prNumber}/files`);
  }

  /**
   * Merge pull request
   */
  async mergePullRequest(
    fullName: string, 
    prNumber: number, 
    options: {
      commit_title?: string;
      commit_message?: string;
      merge_method?: 'merge' | 'squash' | 'rebase';
    } = {}
  ): Promise<any> {
    return this.makeRequest(`/repos/${fullName}/pulls/${prNumber}/merge`, {
      method: 'PUT',
      body: JSON.stringify({
        merge_method: 'merge',
        ...options
      })
    });
  }

  // ========================================================================
  // WEBHOOK OPERATIONS
  // ========================================================================

  /**
   * Get repository webhooks
   */
  async getWebhooks(fullName: string): Promise<any[]> {
    return this.makeRequest(`/repos/${fullName}/hooks`);
  }

  /**
   * Create webhook
   */
  async createWebhook(
    fullName: string,
    config: {
      url: string;
      secret?: string;
      events?: string[];
      active?: boolean;
    }
  ): Promise<any> {
    const webhookData = {
      name: 'web',
      active: config.active ?? true,
      events: config.events || ['push', 'pull_request'],
      config: {
        url: config.url,
        content_type: 'json',
        ...(config.secret && { secret: config.secret })
      }
    };

    return this.makeRequest(`/repos/${fullName}/hooks`, {
      method: 'POST',
      body: JSON.stringify(webhookData)
    });
  }

  /**
   * Update webhook
   */
  async updateWebhook(
    fullName: string,
    hookId: number,
    config: {
      url?: string;
      secret?: string;
      events?: string[];
      active?: boolean;
    }
  ): Promise<any> {
    const webhookData: any = {};
    
    if (config.active !== undefined) {
      webhookData.active = config.active;
    }
    
    if (config.events) {
      webhookData.events = config.events;
    }
    
    if (config.url || config.secret) {
      webhookData.config = {};
      if (config.url) webhookData.config.url = config.url;
      if (config.secret) webhookData.config.secret = config.secret;
    }

    return this.makeRequest(`/repos/${fullName}/hooks/${hookId}`, {
      method: 'PATCH',
      body: JSON.stringify(webhookData)
    });
  }

  /**
   * Delete webhook
   */
  async deleteWebhook(fullName: string, hookId: number): Promise<void> {
    await this.makeRequest(`/repos/${fullName}/hooks/${hookId}`, {
      method: 'DELETE'
    });
  }

  /**
   * Test webhook
   */
  async testWebhook(fullName: string, hookId: number): Promise<any> {
    return this.makeRequest(`/repos/${fullName}/hooks/${hookId}/tests`, {
      method: 'POST'
    });
  }

  // ========================================================================
  // ISSUE OPERATIONS
  // ========================================================================

  /**
   * Get repository issues
   */
  async getIssues(
    fullName: string,
    state: 'open' | 'closed' | 'all' = 'open',
    labels?: string[]
  ): Promise<any[]> {
    const params = new URLSearchParams({
      state,
      sort: 'updated',
      direction: 'desc',
      per_page: '50'
    });

    if (labels && labels.length > 0) {
      params.append('labels', labels.join(','));
    }

    return this.makeRequest(`/repos/${fullName}/issues?${params}`);
  }

  /**
   * Create issue
   */
  async createIssue(
    fullName: string,
    title: string,
    body?: string,
    labels?: string[],
    assignees?: string[]
  ): Promise<any> {
    const issueData: any = { title };
    if (body) issueData.body = body;
    if (labels) issueData.labels = labels;
    if (assignees) issueData.assignees = assignees;

    return this.makeRequest(`/repos/${fullName}/issues`, {
      method: 'POST',
      body: JSON.stringify(issueData)
    });
  }

  // ========================================================================
  // CHECK OPERATIONS
  // ========================================================================

  /**
   * Get check runs for a commit
   */
  async getCheckRuns(fullName: string, ref: string): Promise<any> {
    return this.makeRequest(`/repos/${fullName}/commits/${ref}/check-runs`);
  }

  /**
   * Get check suites for a commit
   */
  async getCheckSuites(fullName: string, ref: string): Promise<any> {
    return this.makeRequest(`/repos/${fullName}/commits/${ref}/check-suites`);
  }

  /**
   * Get combined status for a commit
   */
  async getCombinedStatus(fullName: string, ref: string): Promise<any> {
    return this.makeRequest(`/repos/${fullName}/commits/${ref}/status`);
  }

  // ========================================================================
  // UTILITY METHODS
  // ========================================================================

  /**
   * Check if repository exists and is accessible
   */
  async checkRepositoryAccess(fullName: string): Promise<boolean> {
    try {
      await this.getRepository(fullName);
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Get repository languages
   */
  async getRepositoryLanguages(fullName: string): Promise<Record<string, number>> {
    return this.makeRequest(`/repos/${fullName}/languages`);
  }

  /**
   * Get repository topics
   */
  async getRepositoryTopics(fullName: string): Promise<string[]> {
    const response = await this.makeRequest(`/repos/${fullName}/topics`, {
      headers: {
        'Accept': 'application/vnd.github.mercy-preview+json'
      }
    });
    return response.names || [];
  }

  /**
   * Get repository collaborators
   */
  async getCollaborators(fullName: string): Promise<any[]> {
    return this.makeRequest(`/repos/${fullName}/collaborators`);
  }

  /**
   * Get rate limit status
   */
  async getRateLimit(): Promise<any> {
    return this.makeRequest('/rate_limit');
  }

  /**
   * Parse GitHub URL to extract owner and repo
   */
  parseGitHubUrl(url: string): { owner: string; repo: string } | null {
    const patterns = [
      /github\.com\/([^\/]+)\/([^\/]+)/,
      /github\.com\/([^\/]+)\/([^\/]+)\.git/,
      /git@github\.com:([^\/]+)\/([^\/]+)\.git/
    ];

    for (const pattern of patterns) {
      const match = url.match(pattern);
      if (match) {
        return {
          owner: match[1],
          repo: match[2].replace(/\.git$/, '')
        };
      }
    }

    return null;
  }

  /**
   * Format repository full name
   */
  formatFullName(owner: string, repo: string): string {
    return `${owner}/${repo}`;
  }
}

// ============================================================================
// EXPORT SINGLETON INSTANCE
// ============================================================================

export const githubService = new GitHubService();
